File: two-sum/twoSum.js
// O(n)
function twoNumberSum(array, targetSum) {
  const previousValues = {};
  let i = 0;
  while (i < array.length) {
    const potentialMatch = targetSum - array[i];
    const previousIndex = previousValues[potentialMatch];
    if (previousIndex != null) {
      return [array[previousIndex], array[i]]; // remove "array[" "]" if you want to get index only
    } else {
      previousValues[array[i]] = i;
    }
    i++;
  }
  return [];
}

console.log(twoNumberSum([3, 5, 2, -4, 8, 11], 19));

module.exports = twoNumberSum;

File: two-sum/twoSum.py
# O(n)
def twoNumberSum(arr, target):
    map = {}
    i = 0
    while i < len(arr):
        potentialMatch = target - arr[i]
        previousIndex = map.get(potentialMatch)
        if previousIndex != None:
            return [previousIndex, i] # Add array[ ] if you want to return numbers
        else:
            map[arr[i]] = i
        i += 1
    return []


print(twoNumberSum([3, 5, 2, -4, 8, 11], 19))
